src/plans/first_order_plan.jl

```diff
@@ -1,224 +1,560 @@
 
-Depending on the [`AbstractEvaluationType`](@ref) `T` the gradient can have to forms
+## Keyword arguments
 
-* as a function `(M, p) -> X` that allocates memory for `X`, an [`AllocatingEvaluation`](@ref)
-* as a function `(M, X, p) -> X` that work in place of `X`, an [`InplaceEvaluation`](@ref)
+* `cost = nothing` the cost function `c = f(M,p)`
+* `differential = nothing` the differential `d = df(M, p, X)`
+* `gradient=nothing` the gradient function `g(M, p)` or in-place `g!(M, X, p)`
+* `costgradient = nothing` the combined cost and gradient function `fg(M,p)` or in-place `fg!(M, X, p))`
+* `costdifferential = nothing` the combined cost and differential function  `fdf(M, p, X)`
+* `gradientdifferential=nothing` the combined gradient and differential function `gd(M,p,X)` or its in-place variant `gd!(M; Y, p, X)`
+* `costgradientdifferential = nothing` the single function for all three `fdg(M, p, X)` or `fdg!(M, Y, p, X)`
+$(_var(:Keyword, :evaluation))
 
-# Constructors
-    ManifoldGradientObjective(cost, gradient; evaluation=AllocatingEvaluation())
+Where one can not provide any function twice, that is, e.g. that `costgradient` means neither `cost` nor `gradient` can be provided.
 
 # Used with
 [`gradient_descent`](@ref), [`conjugate_gradient_descent`](@ref), [`quasi_Newton`](@ref)
 """
-struct ManifoldGradientObjective{T<:AbstractEvaluationType,C,G} <:
-       AbstractManifoldGradientObjective{T,C,G}
-    cost::C
-    gradient!!::G
+struct ManifoldFirstOrderObjective{E<:AbstractEvaluationType,F} <:
+       AbstractManifoldFirstOrderObjective{E,F}
+    functions::F
 end
-function ManifoldGradientObjective(
-    cost::C, gradient::G; evaluation::AbstractEvaluationType=AllocatingEvaluation()
-) where {C,G}
-    return ManifoldGradientObjective{typeof(evaluation),C,G}(cost, gradient)
+
+# A Monster constructor
+function ManifoldFirstOrderObjective(;
+    cost=nothing,
+    differential=nothing,
+    gradient=nothing,
+    costgradient=nothing,
+    costdifferential=nothing,
+    gradientdifferential=nothing,
+    costgradientdifferential=nothing,
+    evaluation::E=AllocatingEvaluation(),
+) where {E<:AbstractEvaluationType}
+    # avoid doubles
+    nc = isnothing(cost)
+    nd = isnothing(differential)
+    ng = isnothing(gradient)
+    ncg = isnothing(costgradient)
+    ncd = isnothing(costdifferential)
+    ngd = isnothing(gradientdifferential)
+    ncgd = isnothing(costgradientdifferential)
+
+    t = "provided, one of them twice"
+    (!nc) && (!ncg || !ncd || !ncgd) && error("A cost can not be provided twice.")
+    (!ng) && (!ncg || !ncg || !ncgd) && error("A gradient can not be provided twice.")
+    (!nd) && (!ncd || !ngd || !ncgd) && error("A differential can not be provided twice.")
+    (!ncg) && (!ncd || !ngd || !ncgd) && error("A cost and gradient $t")
+    (!ncd) && (!ngd || !ncgd) && error("Cost and differential $t)")
+    (!ngd) && (!ncgd) && error("Gradient and differential provided twice.")
```

Comment:
What's the problem with providing things twice? I think for example providing cg + cd definitely makes sense.

Reply:
I'd call `cd` if only the cost is requested because it should be cheaper. I'd leave worrying about numerical stability to the user. If the user provides us two ways of computing the same thing, they should expect us to use one that computes the least amount of non-requested output.

Reply:
My personal recommendation for what should be used
Request c: c, cd, cg
Request g: g, cg
Request d: d, cd, g, cg
Request cg: cg, c+g, cd+g
Request cd: cd, c+d, cg

cgd doesn't seem useful actually so I've skipped it.

I wouldn't worry about cache not caching something it potentially could at this point, it can be added later when needed.

Reply:
I think throwing out cdg, allowing for cd+cg and doing the simplest possible approach to caching shouldn't be too difficult.

Reply:
cgd probably makes sense if you _only_ provide that, so it would always be last in your list
I think in practice for now we never request cd, but still good to have that in mind

For caching the simplest possible one is done: Cache only what was asked, ignore rest.

Reply:
cd is requested by some methods from LineSearches.jl, for example Hager-Zhang primarily relies on cd. Very likely some `d` calls could be combined with `c`, for example in `WolfePowellBinaryLinesearchStepsize` or `WolfePowellLinesearchStepsize`.

Reply:
For now there is only a function for cg ( `get_cost_gradient`), since programming all other cases would explode the amount of code – in like for my dispatch I needed some 800 LOC, which is now only 100 or so in the named tuple – but all accessors would introduce maybe 800 or more lines again – so I am not sure how to introduce the combined access variants in a reasonable way. At least the points that I saw, WolfePowell does use differential instead of gradient at least.

Reply:
Oh, I noticed something else that at least for me now (being a total NamedTuple newbie) sounds super complicated in the new idea of allowing everything in several places: constructing the named tuple.

Sure I can remove the few checks atop that currently avoid double costs – but then I have to write 9^9 cases for all the tuples that might appear? That is not yet clear to me how to do that in a finite amount of letters.

Reply:
I've literally written the entire intended logic in a few lines a few posts above. I completely don't understand how you managed to generate so many different combinations.

Reply:
What exactly breaks if you just remove the checks for duplicated costs?

Reply:
If I remove just the checks these comments are on – all that do have duplicates would not define a `nt` (=namedtuple, I was very creative) variable and the following line would error.

The lines after those checks really collect the correct combinations that exist under assuming they are unique and provide a good error message if for example the cost is missing (which one of course also could allow, just having a gd-objective).

The _logic_ is not my problem, the accessors I could maybe write (adapt `get_gradient` to choose the one from your preference list depending on existence).

But NamedTuples are a concept that is nice to use when you have it but creating one with the 9! combinations of tuples we might have here – I am only able to do that in 3 x 9! lines.

Reply:
I think this is now resolved by my last commit.